layout/plan.h

```diff
@@ -0,0 +1,434 @@
+ *
+ *   - IO code in client API, including key-value indices and data objects, and
+ *   - in-storage compute (ISC, aka Function Shipping).
+ *
+ * Interface
+ * ---------
+ *
+ * In the following, "implementation" means generic layout code and particular
+ * layout type code, while "user" denotes a caller of the layout plan interface
+ * (e.g., IO code).
+ *
+ * An access plan is represented by m0_layout_plan.
+ *
+ * The directed graph is not presented as a first-class entity. Instead, a plan
+ * produces, via calls to m0_layout_plan_get(), plops as they become "ready",
+ * in the sense that all their dependencies are satisfied.
+ *
+ * Interaction between the implementation and the user takes form of the loop
+ * (which is, basically, a graph traversal):
+ *
+ *   - the user calls m0_layout_plan_get() to obtain a ready plop. The
+ *     implementation returns a plop, represented by a "sub-class" of
+ *     m0_layout_plop. Some plop fields are set up by the implementation
+ *     before the plop is handed to the user;
+ *
+ *   - the user processes the plop in the appropriate manner, e.g., executes
+ *     read or write operation (which, in turn, might involve layout plans)
+ *     and sets plop fields accordingly to the results of the execution (e.g.,
+ *     points plop to the data read from a cob);
+ *
+ *   - the user calls m0_layout_plop_done() to signal the implementation that
+ *     the plop has been executed. This might make more plops ready, so the
+ *     loop repeats.
+ *
+ * A plop is not necessarily immediately destroyed by the implementation after
+ * the user completes it. The implementation might keep plop alive for some
+ * time, for example, data read from a cob might be needed for a future parity
+ * calculation. When the implementation is done with the plop, it invokes
+ * m0_layout_plop_ops::po_fini() call-back that the user might optionally set
+ * before calling m0_layout_plop_done(). Between the call to
+ * m0_layout_plop_done() and invocation of ->po_fini() call-back the plop (and
+ * the user data buffers referenced by the plop fields) are shared between the
+ * user and the application and the user should neither modify nor free the plop
+ * or its data buffers.
+ *
+ * As an optimisation, implementation deterministically colours plops, in such a
+ * way that a user can improve locality of reference by assigning colours to
+ * processor cores. Colouring is optional and does not affect correctness.
+ *
+ * Let's consider, for example, the i/o read operation. The user calls
+ * m0_layout_plan_build() passing m0_op as argument describing the operation.
+ * After this we can enter the "loop" traversing the graph. m0_layout_plan_get()
+ * is called and returns M0_LAT_READ plop. In this plop user gets cob fid
+ * (at m0_layout_plop::pl_ent) and m0_layout_io_plop::iop_ext describing where
+ * the unit to be read is located at the cob. Now, having this information the
```

Comment:
@nikitadanilov how Function Shipping code (user) can know whether we are going to read the real data or the parity data units? FS needs to know where the data units are located in order to send the computation requests there.

Reply:
@nikitadanilov, what if we introduce an additional plop type - `M0_LAT_MUST_READ` which would indicate that the user must try to read the actual data from the server (data or parity unit) and put it in `iop_data` when calling plop_done()? The simple `M0_LAT_READ` would indicate that the plan is working in normal mode (not the degraded one), so the ISC code could just send the computing request to the server without reading the actual data.

Looks like it would solve the problems for ISC mentioned above.

Reply:
This only works for a particular combination of layout type (parity de-clustered layout type) and particular user (ISC). I don't see how he distinction between `READ` and `MUST_READ` makes sense for an arbitrary user of an arbitrary layout type.

I think the notion of layout plan as a graph (represented by dependencies) is generic enough to capture ISC and other users including normal IO. Let's discuss this on a call. I'll leave two dot-graphs for a normal read and a read with failure as examples.

```digraph plop {
	subgraph cluster_0 {
		style=filled;
		color=lightgrey;
		start;
		getlayout [label="layout-index.GET(fid)"];
		getlayout -> start;
	}
	subgraph cluster_1 {
		style=filled;
		color=lightgrey;
		read0 [label="cob0.read(ext0)"];
		read1 [label="cob1.read(ext1)"];
		read2 [label="cob2.read(ext2)"];
		readout0 [label="buf0.ready"];
		readout1 [label="buf1.ready"];
		readout2 [label="buf2.ready"];
		readout0 -> read0;
		readout1 -> read1;
		readout2 -> read2;
		read0 -> getlayout [style=invis];
		read1 -> getlayout [style=invis];
		read2 -> getlayout [style=invis];
	}
	subgraph cluster_2 {
		style=filled;
		color=lightgrey;
		done
		done -> readout0 [style=invis];
		done -> readout1 [style=invis];
		done -> readout2 [style=invis];
	}
}
```

```
digraph plopfail {
	subgraph cluster_0 {
		style=filled;
		color=lightgrey;
		start;
		getlayout [label="layout-index.GET(fid)"];
		getlayout -> start;
	}
	subgraph cluster_1 {
		style=filled;
		color=lightgrey;
		read0 [label="cob0.read(ext0)" color="red"];
		read1 [label="cob1.read(ext1)"];
		read2 [label="cob2.read(ext2)"];
		readout0 [label="buf0.ready" color="blue"];
		readout1 [label="buf1.ready"];
		readout2 [label="buf2.ready"];
		readout0 -> read0;
		readout1 -> read1;
		readout2 -> read2;
		read0 -> getlayout [style=invis];
		read1 -> getlayout [style=invis];
		read2 -> getlayout [style=invis];
	}
	subgraph cluster_degraded {
		style=filled;
		color=lightgrey;
		readparity [label="cob-parity.read(ext-parity)"];
		readparity -> read0 [style=invis];
		compute [label="erasure(buf1, buf2, buf-parity)"];
		readout0p [label="buf0.ready"];
		compute -> readparity;
		compute -> read1;
		compute -> read2;
		readout0p -> compute;
	}
	subgraph cluster_2 {
		style=filled;
		color=lightgrey;
		done
		done -> readout0p [style=invis];
		done -> readout1 [style=invis];
		done -> readout2 [style=invis];
	}
}
```

Reply:
> I don't see how he distinction between READ and MUST_READ makes sense for an arbitrary user of an arbitrary layout type.

@nikitadanilov, could you please elaborate this statement a bit? For this distinction looks generic enough for ISC needs when a clear indication from the implementation is needed whether the request for the computation can be sent to the server side to be done closer to where the data is stored or not.

Reply:
After discussion with @nikitadanilov there seem to be two major issues with the proposed approach: 1) it won't allow using additional layouts for ISC (like encryption or checksums); 2) it introduces double-reading of the parity group units in the normal i/o path in case of an incidental unforeseeable degraded read (when some READ returns error). Therefore, here is an improved proposition which tackles the above issues and still allows to keep the interface simple for the users:

1. Return READ (instead of MUST_READ) even if there are some FUNctions that must be executed on the read unit data (like checksumming or decompression), but all these functions don't require any other READs in order to get the resulting data for OUT_READ. (Straight line dependency in the graph.) This would tell ISC that the processing of this unit can be done at the server side.

2. Introduce m0_layout_plop_ops::po_update(plop) callback which is called when READ is converted to MUST_READ. This would allow the i/o user to avoid the double-reads on an incidental fallback to degraded read of a parity group. (Note: it is assumed that the OUT_READ plop is returned with parity group granularity and the READ plop is not destroyed until its resulting OUT_READ is ready. This guarantees that po_update(plop) can always be called when needed.)

Reply:
See the proposition in detail at commit https://github.com/Seagate/cortx-motr/pull/411/commits/9fc08d67b01e12f01d6ceb937e3a9753b3286626.

Reply:
As I mentioned during the discussion, it is not clear how the proposal handles the difference between compression (that can be handled by ISC) and client-only encryption (that cannot be handled by ISC).

The only way to define `MUST_READ` precisely and generically, is to define it as `CANNOT_DELEGATE`.  I object to the idea that ISC-specific analysis (that results in the distinction between `READ` and `MUST_READ` plops) should be done by the plan implementation (that is, should be implemented separately by each layout type). This ISC-specific analysis should be done by ISC.

Also, not enough details are provided for the proposed ->po_update() call-back:

    - how the user is supposed to know which plop attributes changed?

    - "The implementation invokes this for a still not complete plop..."---what if the user already completed the `READ` plop during normal IO and already freed the buffer? More generally, what are the ownership and liveness assumptions for plops and buffers in this proposal?